pkg/driver/node.go

```diff
@@ -799,3 +810,71 @@ func collectMountOptions(fsType string, mntFlags []string) []string {
 	}
 	return options
 }
+
+// Struct for JSON patch operations
+type JSONPatch struct {
+	OP    string      `json:"op,omitempty"`
+	Path  string      `json:"path,omitempty"`
+	Value interface{} `json:"value"`
+}
+
+// removeNotReadyTaint removes the taint ebs.csi.aws.com/agent-not-ready from the local node
+// This taint can be optionally applied by users to prevent startup race conditions such as
+// TODO FIXME
+func removeNotReadyTaint(k8sClient cloud.KubernetesAPIClient) error {
+	nodeName := os.Getenv("CSI_NODE_NAME")
+	if nodeName == "" {
+		klog.V(4).InfoS("CSI_NODE_NAME missing, skipping taint removal")
+		return nil
+	}
+
+	clientset, err := k8sClient()
+	if err != nil {
+		klog.V(4).InfoS("Failed to communicate with k8s API, skipping taint removal")
+		return nil //lint:ignore nilerr Failing to communicate with k8s API is a soft failure
+	}
+
+	node, err := clientset.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	var taintsToKeep []corev1.Taint
+	for _, taint := range node.Spec.Taints {
+		if taint.Key != AgentNotReadyNodeTaintKey {
+			taintsToKeep = append(taintsToKeep, taint)
+		} else {
+			klog.V(4).InfoS("Queued taint for removal", "key", taint.Key, "effect", taint.Effect)
+		}
+	}
+
+	if len(taintsToKeep) == len(node.Spec.Taints) {
+		klog.V(4).InfoS("No taints to remove on node, skipping taint removal")
+		return nil
+	}
+
+	patchRemoveTaints := []JSONPatch{
```

Comment:
@torredil  mentioned this, but can we use https://pkg.go.dev/k8s.io/kubernetes/pkg/util/taints#RemoveTaint instead of JSON patch?

Reply:
I originally thought that was possible, but sadly, not really:

Firstly, `RemoveTaint` requires a specific effect, so we'd have to repeat it for all three Effects (and the code wouldn't be very future proof, we'd have to update it whenever a new taint effect is added).

Secondly, and more importantly, `RemoveTaint` only updates the local representation of the node, it doesn't actually make any calls to the k8s API. Thus, you then need to update the entire node object, which (1) is way less efficient than just patching the taints and (2) possibly even introduces a race condition (if the node is updated between us downloading it and attempting to update it.

Reply:
Are there concerns with this implementation?

```
node.Spec.Taints = taintsToKeep
err = clientset.CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
if err != nil {
	return err
}
return nil
```

Reply:
Yes, that implementation would:

> introduces a race condition (if the node is modified by something else between us downloading it and attempting to update it)

It's also just way less efficient to update the entire node object rather than patch the specific part we care about.

Reply:
> (2) possibly even introduces a race condition (if the node is modified by something else between us downloading it and attempting to update it).

K8s requires that objects submitted in update requests contain a `resourceVersion`, the K8s API server verifies this field and will reject the request if there is a conflict which should prevent race conditions.

>  (1) is way less efficient than just patching the taints.

👍 

/lgtm

Reply:
Instead of creating the patch manually, can we `DeepCopy` the Node object, delete the taint, and use `strategicpatch.CreateTwoWayMergePatch` to create the patch data?